lua/entities/gmod_wire_expression2/core/player.lua

```diff
@@ -696,138 +689,151 @@ e2function entity lastDisconnectedPlayer()
+	if not entry then return end -- Player has never respawned.
+	return entry[key]
+end
+
+e2function number lastDeathTime(entity ply) -- When the player provided last died.
+	return getDeathEntry(ply,"timestamp") or 0
+end
+
+e2function entity lastDeathVictim()
+	return DeathList.last.victim
 end
 
 e2function entity lastDeathInflictor()
-	local Inflictor = DeathList.last.inflictor
-	if not IsValid(Inflictor) then return NULL end
-	return Inflictor
+	return DeathList.last.inflictor
 end
 
 e2function entity lastDeathInflictor(entity ply)
-	if not IsValid(ply) then return NULL end
-	if not ply:IsPlayer() then return NULL end
-	local Inflictor = DeathList[ply:EntIndex()].inflictor
-	if not IsValid(Inflictor) then return NULL end
-	return Inflictor
+	return getDeathEntry(ply,"inflictor") or NULL
 end
 
 e2function entity lastDeathAttacker()
-	local Attacker = DeathList.last.attacker
-	if not IsValid(Attacker) then return NULL end
-	return Attacker
+	return DeathList.last.attacker
 end
 
 e2function entity lastDeathAttacker(entity ply)
-	if not IsValid(ply) then return NULL end
-	if not ply:IsPlayer() then return NULL end
-	local Attacker = DeathList[ply:EntIndex()].attacker
-	if not IsValid(Attacker) then return NULL end
-	return Attacker
+	return getDeathEntry(ply,"attacker") or NULL
 end
 
--- Spawn Functions
+-- Respawn Functions
 e2function number spawnClk()
 	return self.data.runByRespawned and 1 or 0
 end
 
-e2function void runOnSpawn(number activate)
-	if activate ~= 0 then
-		RespawnAlert[self.entity] = true
-	else
-		RespawnAlert[self.entity] = nil
-	end
+e2function void runOnSpawn(number activate) -- If 1, make the chip run on a player RESPAWNING. Not joining.
+	RespawnAlert[self.entity] = active~=0 and true or nil
 end
 
 e2function number lastSpawnTime()
-	local Timestamp = RespawnList.last.timestamp
-	if not IsValid(Timestamp) then return 0 end
-	return Timestamp
+	return RespawnList.last.timestamp or 0
 end
 
-e2function number lastSpawnTime(entity ply) -- returns the last time player provided spawned.
-	if not IsValid(ply) then return 0 end
-	if not ply:IsPlayer() then return 0 end
-	local Timestamp = SpawnList[ply:EntIndex()].timestamp
-	if not IsValid(Timestamp) then return 0 end
-	return Timestamp
+e2function number lastSpawnTime(entity ply) -- When the player provided last respawned. (Not joined)
+	return getRespawnEntry(ply,"timestamp") or 0
 end
 
 e2function entity lastSpawnedPlayer()
-	local Ply = RespawnList.last.ply
-	if not IsValid(Ply) then return NULL end
-	return Ply
+	return RespawnList.last.ply
 end
 --******************************************--
+
+
+-- Destructor to avoid invalid chips being called in hooks.
+-- Moved down here to avoid memory leak with Death/Respawn alerts
+
+-- Maybe another E2Lib / WireLib function could be made for this to be automated?
+registerCallback("destruct",function(self)
+	KeyAlert[self.entity] = nil
+	--Used futher below. Didn't want to create more then one of these per file
```

Comment:
this comment and the one below don't need to explain that they are used further below, since this is the end of the file